Voyage Calculation improvements:

ChewableCPP opened the pull request:
This is in reaction to @paulbert's PR #12 .

I reconsidered my earlier, previously unimplemented idea of iterative swapping, but now I remembered why it wouldn't be optimal. There can be (and even in my sporadic testing I encountered some) situations where a swap does not directly result in a better score, unless the swap is recursive - but there is no way to know for certain when to stop recursively swapping without essentially resorting to brute force.

This PR is a compromise - it adds on a refinement step that performs iterative swapping after the brute force portion.

A further improvement is yet possible by using this refinement step to instead extend the filtered roster and perform the brute force again. This combination could be placed into a loop for further gains. I will probably work on this next time.

IAmPicard commented on native/VoyageCalculator.cpp:
```diff
@@ -306,6 +311,91 @@ void VoyageCalculator::fillSlot(size_t iSlot, unsigned int minScore, size_t minD
+	for (;;) {
+		bool refinementFound = false;
+		
+		auto fUpdateBest = [&,this]{
```
(Nitpick): if we make this lambda return a bool, we can call it in both cases below (slot-slot and slot-unslot); it'll return true if the new calculated score is better (and update bestconsidered as today). It'll save us calculating duration twice and will ensure the assertion for the slot-unslot case;.

ChewableCPP replied:
Hmm... I don't follow.

It actually is being called in both cases below. That's why I quickly pulled this chunk of code out into a lambda (if we're going for code reuse - that sanity check should be its own function though).

The odd thing about it is the extra duration calls. slot-slot doesn't care about the duration estimate at all for determining whether to swap or not so the duration calls are both there purely for logging (there are two calls rather than one to preserve the logging order). There's also an extra call in the slot-unslot case, again to preserve the order. This could be made more efficient, but there's no point since this is not even remotely a bottleneck.

ChewableCPP replied:
Also I notice that the lambda's scope is slightly off - it should be inside the following loop since it's only used there (imo).

ChewableCPP replied:
Actually, I think I'm going to rewrite all this anyway, so it's better to not edit it for now.

ChewableCPP commented:
As promised, I've rewritten the algorithm drawing inspiration from @paulbert's #12 PR.

I've removed the "refine" step I added earlier in this PR and instead add an iterative layer on top of the previous implementation. After every brute force attempt, all crew (per slot) will be re-scored based on the estimate resulting from swapping them into the slot. Then the brute force selection is repeated. If this results in an improvement, the process repeats until no further improvement can be found.

This works surprisingly well, generating significantly better results than previously.

It also seems to support my intuition that brute force can't be avoided.

For my current game voyage, I got the following results:

depth 3: 7.36 hours in 20ms
depth 4: 7.35 hours in 46ms
depth 5: 7.36 hours in 236ms
depth 6: 7.36 hours in 2917ms
depth 7: 7.36 hours in 17342ms

Note that although depth 3 in this case yielded the best answer, depth 4 yielded a slightly worse answer.

Same voyage excepting the crew suggested above:

depth 3: 6.31
depth 4: 6.31
depth 5: 6.31
depth 6: 6.30
depth 7: 6.31

That means there's enough complexity that generates an amount of chaos. With only these two samples, I don't know what effect brute force depth has on average - or if there's something that can be done to address this issue of higher depths sometimes yielding worse answers (besides wrapping another loop over this and using the best of several depths).

Perhaps better heuristics are yet possible, but this seems to be very close to optimal and quite fast at depth 5.

I'll continue to test different depths on at least my own voyages.
